Automattic/jetpack
# AI Assistant: handle network errors

Fixes #30867.

## Proposed changes:

* Add `error_network` event on the `SuggestionsEventSource` class, so anytime the connection to the event source fails, this event gets fired
* Handle the `error_network` event on the suggestions hook to set a friendly message to the user, as well as show the retry option

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* This is tricky to test because we need to force a backend error
* The simplest way to do this is to sandbox your `public-api` and add the following filter on your `0-sandbox.php` file:

```
<?php

function force_disable_openai( $result, $blog_id ) {
	return false;
}

add_filter( 'openai_blog_active', 'force_disable_openai', 10, 2 );
```

* The snippet will disable OpenAI for any request on your sandboxed `public-api`
* While sandboxing the API, build this change and go to the block editor
* Open your browser console to inspect the network activity
* Add an AI Assistant block and run some prompt (example: "10 words related to open source software")
* Watch the network activity and confirm that the request to `/jetpack-ai-query` is failing and returning a `500` response (this is what the snippet above forced the API to do)

<img width="794" alt="Screen Shot 2023-05-24 at 15 22 24" src="https://github.com/Automattic/jetpack/assets/6760046/86662e49-7cda-4fbb-82f7-b3e3f83ae65d">

* On the block:
   * notice that a friendly error message shows up
   * notice that the Retry option shows up

<img width="687" alt="Screen Shot 2023-05-24 at 15 22 51" src="https://github.com/Automattic/jetpack/assets/6760046/c9fd5ff7-aac4-42fb-8117-cfa244bb7a81">

* After testing, don't forget to remove the snippet from your `0-sandbox.php` file
